src/engine/SCons/Node/__init__.py

```diff
@@ -1146,11 +1170,11 @@ def get_binfo(self):
         binfo.bsourcesigs = [s.get_ninfo() for s in binfo.bsources]
 
 
-        binfo.bdepends = self.depends
-        binfo.bdependsigs = [d.get_ninfo() for d in self.depends if d not in ignore_set]
+        binfo.bdepends = [d for d in self.depends if d not in ignore_set]
+        binfo.bdependsigs = [d.get_ninfo() for d in self.depends]
 
-        binfo.bimplicit = self.implicit or []
-        binfo.bimplicitsigs = [i.get_ninfo() for i in binfo.bimplicit if i not in ignore_set]
+        binfo.bimplicit = [i for i in self.implicit or [] if i not in ignore_set]
```

Comment:
Is it really guaranteed that the "or" binds stronger than the "for in" and the "if not in" here? This took me 20min to figure out what's going on... ;)